jwst/tweakreg/tests/test_tweakreg.py

```diff
@@ -51,11 +97,207 @@ def test_write_catalog(dummy_source_catalog, tmp_cwd):
+    n_sources = N_EXAMPLE_SOURCES  # a few more than default minobj
+    rng = np.random.default_rng(26)
+    xs = rng.choice(50, n_sources, replace=False) * 8 + 10
+    ys = rng.choice(50, n_sources, replace=False) * 8 + 10
+    for y, x in zip(ys, xs):
+        m0.data[y-1:y+2, x-1:x+2] = [
+            [0.1, 0.6, 0.1],
+            [0.6, 0.8, 0.6],
+            [0.1, 0.6, 0.1],
+        ]
+
+    m1 = m0.copy()
+    # give each a unique filename
+    m0.meta.filename = 'some_file_0.fits'
+    m1.meta.filename = 'some_file_1.fits'
+    c = ModelContainer([m0, m1])
+    return c
+
+
+@pytest.mark.parametrize("with_shift", [True, False])
+def test_tweakreg_step(example_input, with_shift):
+    """
+    A simplified unit test for basic operation of the TweakRegStep
+    when run with or without a small shift in the input image sources
+    """
+    if with_shift:
+        # shift 9 pixels so that the sources in one of the 2 images
+        # appear at different locations (resulting in a correct wcs update)
+        example_input[1].data[:-9] = example_input[1].data[9:]
+        example_input[1].data[-9:] = BKG_LEVEL
+
+    # assign images to different groups (so they are aligned to each other)
+    example_input[0].meta.group_id = 'a'
+    example_input[1].meta.group_id = 'b'
+
+    # make the step with default arguments
+    step = tweakreg_step.TweakRegStep()
+
+    # run the step on the example input modified above
+    result = step(example_input)
+
+    # check that step completed
+    for model in result:
+        assert model.meta.cal_step.tweakreg == 'COMPLETE'
+
+    # and that the wcses differ by a small amount due to the shift above
+    # by projecting one point through each wcs and comparing the difference
+    abs_delta = abs(result[1].meta.wcs(0, 0)[0] - result[0].meta.wcs(0, 0)[0])
+    if with_shift:
+        assert abs_delta > 1E-5
+    else:
+        assert abs_delta < 1E-12
+
+
+@pytest.fixture()
+def custom_catalog_path(tmp_path):
+    fn = tmp_path / "custom_catalog.ecsv"
+
+    # it's important that the sources here don't match
+    # those added in example_input but conform to the input
+    # shape, wcs, etc used in example_input
+    rng = np.random.default_rng(42)
+    n_sources = N_CUSTOM_SOURCES
+    xs = rng.choice(50, n_sources, replace=False) * 8 + 10
+    ys = rng.choice(50, n_sources, replace=False) * 8 + 10
+    catalog = Table(np.vstack((xs, ys)).T, names=['x', 'y'], dtype=[float, float])
+    catalog.write(fn)
+    return fn
+
+
+@pytest.mark.parametrize(
+    "catfile",
+    ["no_catfile", "valid_catfile", "invalid_catfile", "empty_catfile_row"],
+)
+@pytest.mark.parametrize(
+    "asn",
+    ["no_cat_in_asn", "cat_in_asn", "empty_asn_entry"],
+)
+@pytest.mark.parametrize(
+    "meta",
+    ["no_meta", "cat_in_meta", "empty_meta"],
+)
+@pytest.mark.parametrize("custom", [True, False])
+@pytest.mark.slow
+def test_custom_catalog(custom_catalog_path, example_input, catfile, asn, meta, custom, monkeypatch):
```

Comment:
I think this unit test is well-structured, covers a wide range of scenarios, and ensures that the behavior of `TweakRegStep` with custom catalogs is thoroughly tested. However, I think we would benefit from breaking it down into smaller, more readable functions that handle specific aspects of setting up the environment, running the test with different options, and patching methods for testing.

Reply:
Thanks for bringing this up. Splitting this up does seem like one way to make this more readable. However, I tried splitting this apart and I haven't been able to sort out a way to split this up "cleanly". To provide one example of a complication, the test setup splits the example input images into groups `a` and `b`. These group ids are also used during the test run (to check how many sources are identified in each group). Splitting up the test function would mean that these group ids would need to be either hard coded in both the setup and test run sub-functions or brought out as arguments or global variables. I don't think introducing more arguments (or global variables) would be a step towards improving the readability and having them hard-coded in 2 functions seems more problematic than having them hard-coded in 1 function. I am open to suggestions on how to break this up but at the moment I'm failing to think of a way to break it up that improves the readability.